### PR TITLE
fix(cli): unclear error messages when using `projen new` with `--from`

### PR DESCRIPTION
**Pull request by adamelkhayyat**
Improvement to error handling and error messages thrown in cases where `--from` flag is used (ex: `npx projen new --from [SPECIFIED-NPM-PACKAGE]`).

This PR covers the cases:
* where npm package specified does not exist in the registry and gracefully handles error with informative message to user
* where npm package specified is not jsii module and gracefully handles error with informative message to user
* add additional tests

Fixes https://github.com/projen/projen/issues/3303

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/cli/util.ts [hunk 1]
```diff
@@ -55,6 +55,46 @@ export function renderInstallCommand(dir: string, module: string): string {
...
+        paths: [baseDir],
+      })
+    );
+  } catch (error: unknown) {
+    if (
+      error instanceof Error &&
+      "code" in error &&
+      error.code === "MODULE_NOT_FOUND"
+    ) {
+      return undefined;
```
Comment (mrgrain) [inline src/cli/util.ts:90]:
```suggestion
      // the provided module is not a jsii module
      return undefined;
```

File: src/cli/util.ts [hunk 2]
```diff
@@ -55,6 +55,46 @@ export function renderInstallCommand(dir: string, module: string): string {
   return `npm install --save --save-dev -f --no-package-lock --prefix="${dir}" ${module}`;
 }
 
+export function moduleExists(cwd: string, module: string): boolean {
+  try {
+    logging.info(`Validating external module's (${module}) existence...`);
```
Comment (mrgrain) [inline src/cli/util.ts:60]: I don't think this needs to be visible in the regular logs.
```suggestion
    logging.debug(`Checking if external module '${module}' exists...`);
```

And maybe also add a debug log for the various result cases.

File: src/cli/util.ts [hunk 3]
```diff
@@ -55,6 +55,46 @@ export function renderInstallCommand(dir: string, module: string): string {
...
+    const externalModule = !module.startsWith("./");
+    if (externalModule) {
+      exec(`npm view ${module}`, { cwd, stdio: "ignore" });
+      return true;
+    } else {
+      return fs.existsSync(module);
```
Comment (mrgrain) [inline src/cli/util.ts:66]: What is this case covering? Local file?

File: src/cli/util.ts [hunk 4]
```diff
@@ -55,6 +55,46 @@ export function renderInstallCommand(dir: string, module: string): string {
   return `npm install --save --save-dev -f --no-package-lock --prefix="${dir}" ${module}`;
 }
 
+export function moduleExists(cwd: string, module: string): boolean {
+  try {
+    logging.info(`Validating external module's (${module}) existence...`);
+    const externalModule = !module.startsWith("./");
```
Comment (mrgrain) [inline src/cli/util.ts:61]: This feels like a new restriction. We might have to find out how `npm` does the disambiguation.

File: src/cli/cmds/new.ts [hunk 5]
```diff
@@ -292,15 +298,26 @@ async function initProjectFromModule(baseDir: string, spec: string, args: any) {
...
-  const moduleDir = path.dirname(
-    require.resolve(`${moduleName}/.jsii`, {
-      paths: [baseDir],
-    })
-  );
+  const moduleDir = findJsiiFilePath(baseDir, moduleName);
+
+  if (!moduleDir) {
+    throw new CliError(
+      `Cannot find '${moduleName}/.jsii', ensure this is a jsii module first!`
```
Comment (mrgrain) [inline src/cli/cmds/new.ts:318]: Something like this maybe?
```suggestion
      `Module '${moduleName}' does not look like it is compatible with projen. Reason: Cannot find '${moduleName}/.jsii'. All projen module must be jsii modules!`
```

**Review by mrgrain — CHANGES_REQUESTED** [review]
I'm excited by this! A few minor notes.

File: src/cli/util.ts (hunk 3 above)
Comment (adamelkhayyat) [reply src/cli/util.ts:66]: This covers the case when you reference a local tarball (using `npm pack [dep]`).

File: src/cli/util.ts (hunk 4 above)
Comment (adamelkhayyat) [reply src/cli/util.ts:61]: I looked into this further and I believe if I extend the function to check if the module starts with `./` or `/` (absolute paths) it should cover all cases where we're referencing a local file to install with `npm`.

Example installing a local package `some-local-package`:
- `npm install some-local-package` - ❌ Errors out as it attempts to find the package in registry
- `npm install /some-local-package` - ❌ Errors out as it expects absolute path, incomplete absolute path
- `npm install ./some-local-package` - ✅ reference a file that is in the root of the project
- `npm install --from /Users/adam.el-khayyat/Documents/personal-projects/testing-npm-pack/some-local-package`  - ✅ reference via absolute path

💡 I'd also need to adjust the code to accommodate Windows when checking if the local package/file exists (`\` path separator).

Please challenge my conclusion above if you see anything I might've missed or any flaws and I'll adjust as needed!

Comment (mrgrain) [reply src/cli/util.ts:61]: This works for me:

```
npm install some-local-package.tgz
```

Comment (mrgrain) [reply src/cli/util.ts:61]: I think if we extend this test case: https://github.com/projen/projen/blob/main/test/new.test.ts#L122
to run

- without `./` in front
- resolving to an absolute path

we should have tests for most cases

Comment (adamelkhayyat) [reply src/cli/util.ts:61]: > I think if we extend this test case: https://github.com/projen/projen/blob/main/test/new.test.ts#L122 to run
> 
> * without `./` in front
> * resolving to an absolute path
> 
> we should have tests for most cases

In that case we can check if the `module` ends in `.tgz` (and ensure the tarball exists) and we'll cover all existing test cases.

However, this function will return `false` if the `module` references a local project (example: `/Users/adam.el-khayyat/Documents/projects/my-projen-project`) that isn't a tar file compressed and cause a misleading error that the module doesn't exist.

Comment (adamelkhayyat) [reply src/cli/util.ts]: I've extended the function in https://github.com/projen/projen/pull/3385/commits/00ddc146dc9d7018e24475b082cf29e4f75d6da4 to check all conditions to determine if the module is a local one and changed one test based on your comment, let me know if this covers it well enough!

Comment (mrgrain) [reply src/cli/util.ts]: @adamelkhayyat I'd prefer if we have dedicated tests for all 3 cases, i.e. the existing one plus two new variations.

Comment (adamelkhayyat) [reply src/cli/util.ts]: Got it, wanted to confirm the approach first and then write the additional tests.

I'll try to get to this sometime this weekend!

Comment (adamelkhayyat) [reply src/cli/util.ts]: @mrgrain, I've added tests in https://github.com/projen/projen/pull/3385/commits/452d9ef840f56db1a970e63914845b4a644c0661, please let me know what you think!

Comment (adamelkhayyat) [reply src/cli/util.ts]: Note: absolute path test is failing on CI, I'll try to address this sometime this coming week.

File: test/new.test.ts [hunk 6]
```diff
@@ -70,56 +70,105 @@ test("projen new --from external", () => {
...
+
     execProjenCLI(projectdir, [
       "new",
       "--from",
-      "@pepperize/projen-awscdk-app-ts@0.0.333",
+      `${resolve(
+        __dirname,
+        "..",
+        "pepperize-projen-awscdk-app-ts-0.0.333.tgz"
+      )}`,
```
Comment (mrgrain) [inline test/new.test.ts:151]: This should do it
```suggestion
      `${resolve(
        projectdir,
        "pepperize-projen-awscdk-app-ts-0.0.333.tgz"
      )}`,
```

**Comment by mrgrain** [comment]
Thanks @adamelkhayyat I think I can take it from here!

**Review by mrgrain — APPROVED** [review]

**Review by mrgrain — APPROVED** [review]

**Pull request merged**